### DISCUSSION TITLE
UpdateSubresource issues.

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#297] [Q&A]
I'm updating a region of a texture using UpdateSubresource. I've had this code working in SharpDX but I can't get it to work in Vortice. If I pass pitch values the call to UpdateSubresource crashes with the following exception...

> Exception thrown at 0x00007FFC6F367720 (nvwgf2umx_cfg.dll) in DemoSWRenderer.exe: 0xC0000005: Access violation reading location 0x000000EB89E40028.

If I pass no pitch values as in the following code, the texture region written to, but it's corrupted - as if the data is misaligned..
```
    // update the resource
    Gfx.Device3d.DxDevice.ImmediateContext.UpdateSubresource(
        value : dataBox,
        resource: texture.DxTexture2d,
        0,
        0,
        0,
        destRegion
        );
```
![UpdateSubresource](https://user-images.githubusercontent.com/4374176/172149215-b3e71f7c-c057-4306-ab5c-46295d907a92.png)

Here's the state of the arguments in the following code...

texture
    Format = R8G8B8A8_UNorm
    Size = 1802 x 956
    MSCount = 1
    
x,y,w,h = 0,0,50,50

data
    Color8[2500] // data is good upon inspection

Color8...
```    
[StructLayout(LayoutKind.Explicit)]
public struct Color8 : IEquatable<Color8>
{
    [FieldOffset(0)]
    public byte R;
    [FieldOffset(1)]
    public byte G;
    [FieldOffset(2)]
    public byte B;
    [FieldOffset(3)]
    public byte A;
```  
    
Texture writing function ...
```
   public static class TextureWriter2
    {
        public static void Write<T>(Texture texture, T[] data, int x, int y, int w, int h) where T : unmanaged
        {
            // check that the texture is not null
            texture = texture ?? throw new Exception("TextureWriter2.Write | parameter: texture is null");

            int x1 = x;
            int x2 = x + w;
            int y1 = y;
            int y2 = y + h;
            

            // check bounds
            if (x1 < 0 || x2 < x1 || x2 > texture.Width ||
                y1 < 0 || y2 < y1 || y2 > texture.Height)
            {
#if DEBUG
                Debug.LogMessage("TextureWriter2.Write | paramter: region is invalid");
#endif
                return;
            }

            // open stream
            Vortice.DataStream dataStream = new Vortice.DataStream(
                texture.BytesPerPixel * w * h,
                true, true);
            dataStream.WriteRange<T>(data);
            dataStream.Position = 0;

            // form databox
            int rPitch = texture.BytesPerPixel * texture.Width;
            int dPitch = texture.BytesPerPixel * texture.Width * texture.Height;
            Vortice.DataBox dataBox = new Vortice.DataBox(
                dataStream.PositionPointer, rPitch, dPitch);

            // copy cpu data to gpu
            Vortice.Mathematics.Box destRegion = new Vortice.Mathematics.Box
                (
                left: x1,
                top: y1,
                front: 0,
                right: x2,
                bottom: y2,
                back: 1
                );

            // update the resource
            Gfx.Device3d.DxDevice.ImmediateContext.UpdateSubresource(
                value : dataBox,
                resource: texture.DxTexture2d,
                0,
                0,
                0,
                destRegion
                );
            dataStream.Dispose();
        }
    }
```

**Comment by amerkoleci** [comment]
UpdateSubresource had source and destination inverted, Vortice follow DirectX conventions, I'm not backward compatible with SharpDX

**Reply by Gavin-Williams** [reply]
I can't have that part of it wrong though, it looks right to me. And the texture is being updated.

**Reply by amerkoleci** [reply]
I see, can you share a small example so I can try it?

Thanks

**Reply by Gavin-Williams** [reply]
[DemoVorticeUpdateSubresource.zip](https://github.com/amerkoleci/Vortice.Windows/files/8849923/DemoVorticeUpdateSubresource.zip)

**Reply by Gavin-Williams** [reply]
I created a stripped down engine for demonstrating the Vortice code that I'm using. That will help with future issues as well, where I don't want my own abstractions getting in the way.

**Comment by h1cks** [comment]
I use this in my code all the time and it works with Vortice


```
t2d_ = d3dDevice.CreateTexture2D(textureDesc); /// this creates a texture and populates it, we need to create a blank one then populate it.                

 var pixelData_ = BitmapHelper.GetPixelData(bsource);

SwapARGBtoABGR(ref pixelData_);

for (int i = 0; i < pixelData_.Length; i++)
{
buffer_.Write(pixelData_[i]);
}

buffer_.Position = 0;
MappedSubresource box_ = new MappedSubresource(buffer_.PositionPointer, stride_, 0);

d3dDevice.ImmediateContext.UpdateSubresource(box_, t2d_, ID3D11Resource.CalculateSubResourceIndex(0, 0, CountMips(bsource.Size.Width, bsource.Size.Height)));
```

Forgive some names, the code was a port from Sharpdx.

**Reply by amerkoleci** [reply]
Here I showcase how to update texture with new data:
https://github.com/amerkoleci/Vortice.Windows.Samples/blob/main/src/Direct3D11/DrawTexturedCube/Program.cs#L120

**Reply by amerkoleci** [reply]
Latest github has WriteTexture as well:
https://github.com/amerkoleci/Vortice.Windows/blob/f56e69522f1d7abad7c943f9a1197a00e0a7f9df/src/Vortice.Direct3D11/ID3D11DeviceContext.cs#L1904

**Reply by Gavin-Williams** [reply]
@amerkoleci  I think you have the wrong calculation for slicePitch (which is now called depthPitch).

https://docs.microsoft.com/en-us/windows/win32/api/d3d11/nf-d3d11-id3d11devicecontext-updatesubresource

In your sample code
  rowPitch = 16
  slicePitch = 4
  
From the docs page...

for a 2x3 texture (16 bytes per pixel):
Source Row Pitch = 16 * 2 = 32
Source Depth Pitch = 16 * 2 * 3 = 96

And so, in FormatHelper, this is the wrong calculation as well..
```slicePitch = height;```
This should be...
```depthPitch = sizeInBytes * width * height; // I would suggest using the DirectX term of depth instead of slice here. Just to be consistent.``` 

And what's this? ...
```rowPitch = (width * bpp + 7) / 8;```

That's not correct - edit - ok, you use bits here? Not sure if that's necessary, and a bit confusing. I would just use...
``` rowPitch = sizeInBytes * width;```

**Reply by amerkoleci** [reply]
Nah,
RowPitch is pixel format size * width
DepthPitch/SlicePitch = RowPitch * height 

Those are ported for DirectXTex

**Reply by Gavin-Williams** [reply]
But you have slicePitch = height (FomatHelper, Line 280)

**Reply by Gavin-Williams** [reply]
Look at your slicePitch value .. 4 bytes? How can the image be 4 bytes? Slice/depth pitch is the size in bytes from depth 0 to depth 1. That's the size of the texture in bytes (for 2d texture).

**Reply by Gavin-Williams** [reply]
Here's a link that makes it very clear how to compute row pitch and depth pitch.
https://docs.microsoft.com/en-us/windows/win32/api/d3d11/nf-d3d11-id3d11devicecontext-updatesubresource#:~:text=To%20calculate%20the,of%20rows%20(height)%5D

**Reply by Gavin-Williams** [reply]
@h1cks Thanks, that method works fine.

**Comment by Gavin-Williams** [comment]
@amerkoleci If you look at the following code in DirectXTex...
https://github.com/microsoft/DirectXTex/blob/9c72f2c6cdbe3cc9e33ec55f1b15cee356c4ecf6/DDSTextureLoader/DDSTextureLoader11.cpp#L605
```
        else
        {
            const size_t bpp = BitsPerPixel(fmt);
            if (!bpp)
                return E_INVALIDARG;

            rowBytes = (uint64_t(width) * bpp + 7u) / 8u; // round up to nearest byte
            numRows = uint64_t(height);
            numBytes = rowBytes * height;
        }
```
and compare it to Vortice...
```
        else
        {
            int bpp = SizeOfInBits(format);
            rowPitch = (width * bpp + 7) / 8; // round up to nearest byte
            slicePitch = height;
            sizeInBytes = rowPitch * height;
        }
```
It looks live you've transcribed numRows incorrectly to slicePitch - they are not the same.

numRows = y resolution
slicePitch = bytesPerPixel * y-resolution * x-resolution
or
slicePitch = rowBytes * height

**Reply by amerkoleci** [reply]
Patience, I'm on it :) currently FIXING it

**Reply by amerkoleci** [reply]
FIX is here: https://github.com/amerkoleci/Vortice.Windows/commit/117ef31eed16aa51651fae6f77fa54e60b724e12
I plan to release new stable nuget release today, stay tuned!

**Reply by amerkoleci** [reply]
Stable 2.1.18 has been released.

**Comment by Gavin-Williams** [comment]
So it turns out there is still an issue. I can use the DeviceContext.WriteTexture method to update a texture. But if I try to use UpdateSubresource method directly, I get the following error...
> Exception thrown at 0x00007FFC6F367720 (nvwgf2umx_cfg.dll) in DemoVorticeUpdateSubresource.exe: 0xC0000005: Access violation reading location 0x0000007643F800A0.

(note: I also saw this crash when I tried creating a texture with initial data of size that didn't match the texture size - maybe a clue).

Here's the function that fails...
```
        public unsafe static void Write<T>(ID3D11Texture2D texture, T[] data, int x, int y, int w, int h) where T : unmanaged
        {
            // open stream
            int bytesPerPixel = SimpleEngine.FormatHelper.BytesPerPixel(texture.Description.Format);
            Vortice.DataStream dataStream = new Vortice.DataStream(
                bytesPerPixel * w * h,
                true, true);
            dataStream.WriteRange<T>(data);
            dataStream.Position = 0;

            // form databox
            int dataRPitch = bytesPerPixel * w;
            int dataDPitch = bytesPerPixel * w * h;
            Vortice.DataBox dataBox = new Vortice.DataBox(
                dataStream.BasePointer, dataRPitch, dataDPitch);

            // map/copy cpu data to gpu
            Vortice.Mathematics.Box destRegion = new Vortice.Mathematics.Box
                (
                left: x,
                top: y,
                front: 0,
                right: x+w,
                bottom: y+h,
                back: 1
                );

            // update the resource
            Texture2DDescription desc = texture.Description;
            int rPitch = bytesPerPixel * desc.Width;
            int dPitch = bytesPerPixel * desc.Width * desc.Height;

            // check pitch values against Vortice helper - looks good
            { // unused
                Vortice.DXGI.FormatHelper.GetSurfaceInfo(
                    desc.Format, desc.Width, desc.Height,
                    out int rowPitch, out int slicePitch, out int rowCount);
            }

            Engine.DxDevice.ImmediateContext.UpdateSubresource(
                value: dataBox,
                resource: texture,
                subresource: 0,
                rowPitch: rPitch,
                depthPitch: dPitch,
                region: destRegion
                );
            dataStream.Dispose();
        }
```
And here's some calling code, which should give a black to blue transition. 
*** **Switch the write function at the end to see the failure** ***
```
    int pX = 0;
    int pY = 0;
    int pW = 100;
    int pH = 100;
    
    ...
    
    private void UpdateTexture()
    {
        // build new image data
        Color[] dataAsStruct = new Color[pW * pH];
        for (int y = 0; y < pH; y++)
            for (int x = 0; x < pW; x++)
                dataAsStruct[y * pW + x] = new Color(
                    r: 0,
                    g: 0,
                    b: (byte)(x / (pW - 1f) * 255),
                    a: 255);
        ReadOnlySpan<Color> pixels = new ReadOnlySpan<Color>(dataAsStruct);
        
        Engine.DxDeviceContext.WriteTexture(texture, 0, 0, pixels);
        //TextureWriter.Write<Color>(texture, dataAsStruct, pX, pY, pW, pH);
    }
```

Update : The following method based upon @h1cks code suggestion above does work...
```
        public static void WriteUsingMap<T>(ID3D11Texture2D texture, T[] data, int x, int y, int w, int h) where T : unmanaged
        {
            int bytesPerPixel = SimpleEngine.FormatHelper.BytesPerPixel(texture.Description.Format);
            Vortice.DataStream dataStream = new(bytesPerPixel * w * h, true, true);
            dataStream.WriteRange<T>(data);
            dataStream.Position = 0;

            // define destination region
            Vortice.Mathematics.Box destRegion = new Vortice.Mathematics.Box
                (
                left: x,
                top: y,
                front: 0,
                right: x + w,
                bottom: y + h,
                back: 1
                );

            int dataRPitch = bytesPerPixel * w;
            int dataDPitch = bytesPerPixel * w * h;
            MappedSubresource mappedRes = new (dataStream.PositionPointer, dataRPitch, dataDPitch);
            int subResIndex = texture.CalculateSubResourceIndex(0, 0, out int mipSize);
            Engine.DxDeviceContext.UpdateSubresource(mappedRes, texture, subResIndex, destRegion);
            dataStream.Dispose();
        }
```
